### ISSUE TITLE
The weight filed of `Chain` is buggy

### ISSUE THREAD

**Issue by xukai92** [TuringLang/Turing.jl#406]
See discussion in #405

**Comment by xukai92** [comment]
NOTE:

In `io.jl`

```julia
function Base.getindex(c::Chain, v::Symbol)
  # This strange implementation is mostly to keep backward compatability.
  #  Needs some refactoring a better format for storing results is available.
  if v == :logevidence
    log(c.weight)
  elseif v==:samples
    c.value2
  elseif v==:logweights
    c[:lp]
  else
    map((s)->Base.getindex(s, v), c.value2)
  end
end
```

**Comment by xukai92** [comment]
To summarize the discussion in #405,

For `Chain`, we store original `Sample` in `Chain.vals2` so we can get all the weights by `weights = [results.value2[i].weight for i in range(1, length(results))]`.

However in current `bnp-project` branch, I guess due to the limitation of `Chain` that it cannot store varying dimensional variables, the `Vector{Sample}` is directly returned. So `weights = [results[i].weight for i in range(1, length(results))]` is needed to extract weights.

@emilemathieu does it sound correct to you?

**Comment by xukai92** [comment]
Forget to @ae-foster

**Comment by emilemathieu** [comment]
@xukai92 I believe the issue is mostly that the unbiased estimator in importance sampling or SMC is the __weighted__ sum over samples. Therefore, either a resampling step is needed at the end so that the weights are equals (thus useless) or a simple access to the weights is needed.

**Comment by xukai92** [comment]
Yes, I understand this. I'm talking about how to give a simple & consistent access to the weights. There is such an interface but it seems not consistent for `Chain` type, which is returned after inference in the master, and `Vector{Sample}` type, which is used in the bnp branch (?).

**Comment by yebai** [comment]
This should be fixed by #793.
